### ISSUE TITLE
JS config never gets garbage collected.

### ISSUE THREAD

**Issue by SevInf** [davidtheclark/cosmiconfig#90]
I discovered this issue while trying to catch a memory leak in our build tool. The way cosmiconfig loads '.js' config prevents it from ever getting garbage collected, even if userland code no longer needs. `require-from-string` package [creates a new `Module` record](https://github.com/floatdrop/require-from-string/blob/master/index.js#L24) which then permanently stored in memory. You can see it with this sample:

```js
const cosmiconfig = require('cosmiconfig');

const loadRC = require.resolve('cosmiconfig/lib/loadRc.js');

function loadOnce() {
    const explorer = cosmiconfig('example');
    return explorer.load(__dirname)
        .then(() => console.log(require.cache[loadRC].children.length))
}

let count = 1000;
function loop() {
    if (count > 0) {
        count--;
        loadOnce().then(loop);
    }
}

loop();
```

Provided there is a non-empty `example.config.js` nearby, you should see that `require.cache[loadRC].children.length` grows by 1 after every load and never gets cleaned up.

**Comment by davidtheclark** [comment]
That's a bummer and also very interesting, @SevInf. Thanks for investigating it. Do you have any recommendations that might fix it?

**Comment by sudo-suhas** [comment]
> Thanks for investigating it

+1 from me as well.

Is this a bug in `require-from-string` or is the problem related to how `cosmiconfig` uses it?

**Comment by SevInf** [comment]
@davidtheclark it looks like replacing `require-from-string` with something like [`import-fresh`](https://github.com/sindresorhus/import-fresh) fixes it. I can send pull request tomorrow.

@sudo-suhas from `require-from-string` point of view, I think, this behavior is correct, it just might not align that well with `cosmiconfig`.

**Comment by sudo-suhas** [comment]
Thanks @SevInf. Isn't [`import-from`](https://github.com/sindresorhus/import from) more suited to the purpose than `import-fresh`?

**Comment by sudo-suhas** [comment]
Sorry. Closed by mistake.

**Comment by SevInf** [comment]
@sudo-suhas I think by the time it gets to [`loadJs`](https://github.com/davidtheclark/cosmiconfig/blob/master/src/loadJs.js#L8) cosmiconfig have already figured out the absolute path to the file, so `import-from` won't add anything new to a plain `require` here. 

I assumed `require-from-string` is used instead of `require` to bypass `require.cache`. Is it true? If so, `import-fresh` will preserve this behavior.

**Comment by sudo-suhas** [comment]
Honestly not too sure. Perhaps @davidtheclark can clarify. AFAIK there is no intent to bypass `require.cache`. Also see https://github.com/sindresorhus/import-from/issues/4#issuecomment-304430710.

**Comment by davidtheclark** [comment]
@SevInf @sudo-suhas: For a little background you can read https://github.com/davidtheclark/cosmiconfig/issues/21 and https://github.com/davidtheclark/cosmiconfig/pull/51. 

The summary: We were using require-from-string because it allowed us to avoid reading the file multiple times and to stay asynchronous. Those are the things we would sacrifice, I think, if we switched to require. What do you think about the trade-offs?

**Comment by SevInf** [comment]
Ok, so far I've submitted a patch to `require-from-string`, doing the cleanup: floatdrop/require-from-string#9.

I'm still not entirely sure it belongs there, let's see what maintainers say. But if it won't be merged, I think we still can continue to use `require-from-string` and do this cleanup in `cosmiconfig` instead.

**Comment by sudo-suhas** [comment]
I tried to upgrade deps so that this issue would be fixed. However, faced an issue due to jest + `require-from-string` which is failing a lot of tests. Have created a PR floatdrop/require-from-string#10 which will fix the issue.

**Comment by sudo-suhas** [comment]
Fixed and released in 3.0.1 🎉
